docs/design_notes/jaxlib_versioning.md

```diff
@@ -0,0 +1,153 @@
+level.
+
+`jaxlib` is built using Bazel out of the `jax` repository. The pieces of
+`jaxlib` from the XLA repository are incorporated into the build
+[as a Bazel submodule](https://github.com/google/jax/blob/main/WORKSPACE).
+To update the version of XLA used during the build, one must update the pinned
+version in the Bazel `WORKSPACE`. This is done manually on an
+as-needed basis, but can be overriden on a build-by-build basis.
+
+
+## How do we make changes across the `jax` and `jaxlib` boundary?
+
+Changes to `jaxlib` must usually be made in a backwards-compatible
+way. That is, changes to `jaxlib` must maintain API compatibility with both
+current and future `jax` releases. For example, it is often
+safe to add a new function to `jaxlib`, but unsafe to remove an existing
+function or to change its signature without potentially breaking older `jax`
+versions.
+
+Even if we only make backwards-compatible changes to `jaxlib`, we need to be
+careful in what `jax` assumes of the API provided by `jaxlib`. To determine
+whether it is safe for `jax` to assume that feature exists in the `jaxlib` API,
+`jax` uses several version numbers.
+
+`jax` specifies a
+[minimum jaxlib version](https://github.com/google/jax/blob/main/jax/version.py).
+That is, when imported, `jax` checks that the version of the `jaxlib` wheel
+is at least as new as the minimum version. `jax` checks the `jaxlib` version at
+runtime rather than using a `pip` package requirement because we
+[provide separate `jaxlib` wheels](https://github.com/google/jax#installation)
+for a variety of hardware and software versions (e.g, GPU, TPU, etc.), and we
+do not know which is the right choice for any given user.
+
+(In the future, we hope to separate out the hardware-specific pieces of `jaxlib`
+into separate plugins, at which point the minimum version could be expressed as
+a Python package dependency.)
+
+The minimum jaxlib version is a coarse instrument: it only allows us to
+verify that the release number of `jaxlib` is new enough to incorporate a
+particular change. Its primary function is to allow `jax` to drop
+backward-compatibility Python code for older `jaxlib` releases. To drop
+backward-compatibility code, we can increase the minimum jaxlib version and then
+delete any code paths in `jax` that existed to maintain compatibility with older
+`jaxlib` releases. 
+
+However, since the `jax` and `jaxlib` code is split across repositories that
+cannot be updated atomically at the same time, we need to manage compatibility
+at a finer granularity than our release cycle. To manage fine-grained
+compatibility, we have additional versioning that is independent of the `jaxlib`
+release version numbers.
+
+We maintain an additional version number (`_version`) in [`xla_client.py` in
+the XLA repository](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/python/xla_client.py#L45). The idea is that this version number, is defined in `xla/python`
```

Comment:
Possibly link to a release tag rather than `master` here, to avoid the link becoming outdated?

Reply:
Removed the line number instead, which should keep the link reasonably fresh.